packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts

```diff
@@ -0,0 +1,261 @@
+
+/**
+ * Firestore subscriptions contain additional expanded information
+ * on top of the base Stripe.Subscription type
+ */
+export interface FirestoreSubscription extends Stripe.Subscription {
+  customer: string;
+  plan: Stripe.Plan;
+  price: Stripe.Price;
+}
+
+export class PlanCanceller {
+  private config: ConfigType;
+  private firestore: Firestore;
+  private stripeQueue: PQueue;
+  private stripe: Stripe;
+
+  /**
+   * A tool to cancel all subscriptions under a plan
+   * @param priceId A Stripe plan or price ID for which all subscriptions will be cancelled
+   * @param refund If true, all subscriptions will have their last charge reversed to their card, regardless of remaining time
+   * @param prorate If true, all subscriptions cancellations will generate a proration invoice item that credits remaining time
+   * @param excludePlanIds A list of Stripe plan or price ID which if customers have will not be subscribed to the destination plan
+   * @param batchSize Number of subscriptions to fetch from Firestore at a time
+   * @param outputFile A CSV file to output a report of affected subscriptions to
+   * @param stripeHelper An instance of StripeHelper
+   * @param rateLimit A limit for number of stripe requests within the period of 1 second
+   * @param database A reference to the FXA database
+   */
+  constructor(
+    private priceId: string,
+    private refund: boolean,
+    private prorate: boolean,
+    private excludePlanIds: string[],
+    private batchSize: number,
+    private outputFile: string,
+    private stripeHelper: StripeHelper,
+    private database: any,
+    public dryRun: boolean,
+    rateLimit: number
+  ) {
+    this.stripe = this.stripeHelper.stripe;
+
+    const config = Container.get<ConfigType>(AppConfig);
+    this.config = config;
+
+    const firestore = Container.get<Firestore>(AuthFirestore);
+    this.firestore = firestore;
+
+    this.stripeQueue = new PQueue({
+      intervalCap: rateLimit,
+      interval: 1000, // Stripe measures its rate limit per second
+    });
+  }
+
+  /**
+   * Cancel all customer subscriptions in batches
+   */
+  async run() {
+    let startAfter: string | null = null;
+    let hasMore = true;
+
+    while (hasMore) {
+      const subscriptions = await this.fetchSubsBatch(startAfter);
+
+      startAfter = subscriptions.at(-1)?.id as string;
+      if (!startAfter) hasMore = false;
+
+      await Promise.all(
+        subscriptions.map((sub) => this.processSubscription(sub))
+      );
+    }
+  }
+
+  /**
+   * Fetches subscriptions from Firestore paginated by batchSize
+   * @param startAfter ID of the last element of the previous batch for pagination
+   * @returns A list of subscriptions from firestore
+   */
+  async fetchSubsBatch(startAfter: string | null) {
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const subscriptionCollection = `${collectionPrefix}subscriptions`;
```

Comment:
Not strictly necessary. In past scripts had been using other collections so it had more of a point. It could be removed, but also not hindering anything